File: backend/app/api/discussion.py
from flask import Blueprint, request, jsonify, session
from app.models.discussion import Topic, Post, Reply, PostLike, ReplyLike
from app.extensions import db
from app.api import discussion_bp
from sqlalchemy.exc import SQLAlchemyError # type: ignore
from datetime import datetime
from functools import wraps

# 用戶身份驗證裝飾器
def login_required(f):
    @wraps(f)
    def decorated(*args, **kwargs):
        if 'user_id' not in session:
            return jsonify({"message": "請先登入"}), 401
        
        return f(*args, **kwargs)
    return decorated

# 主題相關 API
@discussion_bp.route('/topics', methods=['GET'])
def get_topics():
    """獲取所有討論主題"""
    page = request.args.get('page', 1, type=int)
    per_page = request.args.get('per_page', 10, type=int)
    
    # 置頂主題優先，再按更新時間排序
    query = Topic.query.order_by(Topic.is_pinned.desc(), Topic.updated_at.desc())
    topics_pagination = query.paginate(page=page, per_page=per_page, error_out=False)
    
    result = {
        'topics': [topic.to_dict() for topic in topics_pagination.items],
        'total': topics_pagination.total,
        'pages': topics_pagination.pages,
        'current_page': page
    }
    
    return jsonify(result)


@discussion_bp.route('/topics/<int:topic_id>', methods=['GET'])
def get_topic(topic_id):
    """獲取特定主題詳情"""
    topic = Topic.query.get_or_404(topic_id)
    
    # 增加瀏覽次數
    topic.view_count += 1
    db.session.commit()
    
    return jsonify(topic.to_dict())


@discussion_bp.route('/topics', methods=['POST'])
@login_required
def create_topic():
    """創建新主題"""
    data = request.get_json()
    user_id = session.get('user_id')
    
    if not data.get('title'):
        return jsonify({'error': '主題標題不能為空'}), 400
    
    try:
        topic = Topic(
            title=data.get('title'),
            description=data.get('description', ''),
            creator_id=user_id
        )
        
        db.session.add(topic)
        db.session.commit()
        
        return jsonify(topic.to_dict()), 201
    except SQLAlchemyError as e:
        db.session.rollback()
        return jsonify({'error': f'創建主題失敗: {str(e)}'}), 500


@discussion_bp.route('/topics/<int:topic_id>', methods=['PUT'])
@login_required
def update_topic(topic_id):
    """更新主題資訊"""
    data = request.get_json()
    user_id = session.get('user_id')
    
    topic = Topic.query.get_or_404(topic_id)
    
    # 只有創建者可以更新主題
    if topic.creator_id != user_id:
        return jsonify({'error': '您沒有權限更新此主題'}), 403
    
    try:
        if 'title' in data:
            topic.title = data['title']
        if 'description' in data:
            topic.description = data['description']
            
        db.session.commit()
        return jsonify(topic.to_dict())
    except SQLAlchemyError as e:
        db.session.rollback()
        return jsonify({'error': f'更新主題失敗: {str(e)}'}), 500


@discussion_bp.route('/topics/<int:topic_id>', methods=['DELETE'])
@login_required
def delete_topic(topic_id):
    """刪除主題 (僅限創建者或管理員)"""
    user_id = session.get('user_id')
    
    topic = Topic.query.get_or_404(topic_id)
    
    # 只有創建者或管理員可以刪除主題
    if topic.creator_id != user_id and not session.get('is_admin', False):
        return jsonify({'error': '您沒有權限刪除此主題'}), 403
    
    try:
        db.session.delete(topic)
        db.session.commit()
        return jsonify({'message': '主題已成功刪除'})
    except SQLAlchemyError as e:
        db.session.rollback()
        return jsonify({'error': f'刪除主題失敗: {str(e)}'}), 500


# 文章相關 API

@discussion_bp.route('/topics/<int:topic_id>/posts', methods=['GET'])
def get_posts(topic_id):
    """獲取特定主題下的所有文章"""
    page = request.args.get('page', 1, type=int)
    per_page = request.args.get('per_page', 20, type=int)
    
    # 確認主題存在
    topic = Topic.query.get_or_404(topic_id)
    
    # 置頂文章優先，再按創建時間排序
    query = Post.query.filter_by(topic_id=topic_id).order_by(Post.is_pinned.desc(), Post.created_at.asc())
    posts_pagination = query.paginate(page=page, per_page=per_page, error_out=False)
    
    result = {
        'topic': topic.to_dict(),
        'posts': [post.to_dict() for post in posts_pagination.items],
        'total': posts_pagination.total,
        'pages': posts_pagination.pages,
        'current_page': page
    }
    
    return jsonify(result)


@discussion_bp.route('/topics/<int:topic_id>/posts', methods=['POST'])
@login_required
def create_post(topic_id):
    """在特定主題下發表文章"""
    data = request.get_json()
    user_id = session.get('user_id')
    
    # 確認主題存在
    topic = Topic.query.get_or_404(topic_id)
    
    if not data.get('content'):
        return jsonify({'error': '文章內容不能為空'}), 400
    
    try:
        post = Post(
            topic_id=topic_id,
            author_id=user_id,
            content=data.get('content')
        )
        
        db.session.add(post)
        db.session.commit()
        
        return jsonify(post.to_dict()), 201
    except SQLAlchemyError as e:
        db.session.rollback()
        return jsonify({'error': f'發表文章失敗: {str(e)}'}), 500


@discussion_bp.route('/posts/<int:post_id>', methods=['PUT'])
@login_required
def update_post(post_id):
    """更新文章內容"""
    data = request.get_json()
    user_id = session.get('user_id')
    
    post = Post.query.get_or_404(post_id)
    
    # 只有作者可以更新文章
    if post.author_id != user_id:
        return jsonify({'error': '您沒有權限更新此文章'}), 403
    
    if not data.get('content'):
        return jsonify({'error': '文章內容不能為空'}), 400
    
    try:
        post.content = data.get('content')
        db.session.commit()
        return jsonify(post.to_dict())
    except SQLAlchemyError as e:
        db.session.rollback()
        return jsonify({'error': f'更新文章失敗: {str(e)}'}), 500


@discussion_bp.route('/posts/<int:post_id>', methods=['DELETE'])
@login_required
def delete_post(post_id):
    """刪除文章 (僅限作者或管理員)"""
    user_id = session.get('user_id')
    
    post = Post.query.get_or_404(post_id)
    
    # 只有作者或管理員可以刪除文章
    if post.author_id != user_id and not session.get('is_admin', False):
        return jsonify({'error': '您沒有權限刪除此文章'}), 403
    
    try:
        db.session.delete(post)
        db.session.commit()
        return jsonify({'message': '文章已成功刪除'})
    except SQLAlchemyError as e:
        db.session.rollback()
        return jsonify({'error': f'刪除文章失敗: {str(e)}'}), 500


# 回覆相關 API

@discussion_bp.route('/posts/<int:post_id>/replies', methods=['GET'])
def get_replies(post_id):
    """獲取特定文章的所有回覆"""
    page = request.args.get('page', 1, type=int)
    per_page = request.args.get('per_page', 50, type=int)
    
    # 確認文章存在
    post = Post.query.get_or_404(post_id)
    
    query = Reply.query.filter_by(post_id=post_id).order_by(Reply.created_at.asc())
    replies_pagination = query.paginate(page=page, per_page=per_page, error_out=False)
    
    result = {
        'post': post.to_dict(),
        'replies': [reply.to_dict() for reply in replies_pagination.items],
        'total': replies_pagination.total,
        'pages': replies_pagination.pages,
        'current_page': page
    }
    
    return jsonify(result)


@discussion_bp.route('/posts/<int:post_id>/replies', methods=['POST'])
@login_required
def create_reply(post_id):
    """對特定文章發表回覆"""
    data = request.get_json()
    user_id = session.get('user_id')
    
    # 確認文章存在
    post = Post.query.get_or_404(post_id)
    
    if not data.get('content'):
        return jsonify({'error': '回覆內容不能為空'}), 400
    
    try:
        reply = Reply(
            post_id=post_id,
            author_id=user_id,
            content=data.get('content')
        )
        
        db.session.add(reply)
        db.session.commit()
        
        return jsonify(reply.to_dict()), 201
    except SQLAlchemyError as e:
        db.session.rollback()
        return jsonify({'error': f'發表回覆失敗: {str(e)}'}), 500


@discussion_bp.route('/replies/<int:reply_id>', methods=['PUT'])
@login_required
def update_reply(reply_id):
    """更新回覆內容"""
    data = request.get_json()
    user_id = session.get('user_id')
    
    reply = Reply.query.get_or_404(reply_id)
    
    # 只有作者可以更新回覆
    if reply.author_id != user_id:
        return jsonify({'error': '您沒有權限更新此回覆'}), 403
    
    if not data.get('content'):
        return jsonify({'error': '回覆內容不能為空'}), 400
    
    try:
        reply.content = data.get('content')
        db.session.commit()
        return jsonify(reply.to_dict())
    except SQLAlchemyError as e:
        db.session.rollback()
        return jsonify({'error': f'更新回覆失敗: {str(e)}'}), 500


@discussion_bp.route('/replies/<int:reply_id>', methods=['DELETE'])
@login_required
def delete_reply(reply_id):
    """刪除回覆 (僅限作者或管理員)"""
    user_id = session.get('user_id')
    
    reply = Reply.query.get_or_404(reply_id)
    
    # 只有作者或管理員可以刪除回覆
    if reply.author_id != user_id and not session.get('is_admin', False):
        return jsonify({'error': '您沒有權限刪除此回覆'}), 403
    
    try:
        db.session.delete(reply)
        db.session.commit()
        return jsonify({'message': '回覆已成功刪除'})
    except SQLAlchemyError as e:
        db.session.rollback()
        return jsonify({'error': f'刪除回覆失敗: {str(e)}'}), 500


# 讚相關 API

@discussion_bp.route('/posts/<int:post_id>/like', methods=['POST'])
@login_required
def like_post(post_id):
    """對文章按讚/取消讚"""
    user_id = session.get('user_id')
    
    # 確認文章存在
    post = Post.query.get_or_404(post_id)
    
    # 檢查是否已按讚
    existing_like = PostLike.query.filter_by(post_id=post_id, user_id=user_id).first()
    
    try:
        if existing_like:
            # 取消讚
            db.session.delete(existing_like)
            message = "已取消對文章的讚"
        else:
            # 按讚
            like = PostLike(post_id=post_id, user_id=user_id)
            db.session.add(like)
            message = "已成功對文章按讚"
            
        db.session.commit()
        
        return jsonify({
            'message': message, 
            'likes': [like.user_id for like in post.likes.all()],
            'like_count': post.likes.count()
        })
    except SQLAlchemyError as e:
        db.session.rollback()
        return jsonify({'error': f'操作失敗: {str(e)}'}), 500


@discussion_bp.route('/replies/<int:reply_id>/like', methods=['POST'])
@login_required
def like_reply(reply_id):
    """對回覆按讚/取消讚"""
    user_id = session.get('user_id')
    
    # 確認回覆存在
    reply = Reply.query.get_or_404(reply_id)
    
    # 檢查是否已按讚
    existing_like = ReplyLike.query.filter_by(reply_id=reply_id, user_id=user_id).first()
    
    try:
        if existing_like:
            # 取消讚
            db.session.delete(existing_like)
            message = "已取消對回覆的讚"
        else:
            # 按讚
            like = ReplyLike(reply_id=reply_id, user_id=user_id)
            db.session.add(like)
            message = "已成功對回覆按讚"
            
        db.session.commit()
        
        return jsonify({
            'message': message, 
            'likes': [like.user_id for like in reply.likes.all()],
            'like_count': reply.likes.count()
        })
    except SQLAlchemyError as e:
        db.session.rollback()
        return jsonify({'error': f'操作失敗: {str(e)}'}), 500


# 管理員 API

@discussion_bp.route('/topics/<int:topic_id>/pin', methods=['POST'])
@login_required
def pin_topic(topic_id):
    """置頂/取消置頂主題 (僅限管理員)"""
    if not session.get('is_admin', False):
        return jsonify({'error': '您沒有管理員權限'}), 403
    
    topic = Topic.query.get_or_404(topic_id)
    
    try:
        topic.is_pinned = not topic.is_pinned
        db.session.commit()
        
        status = "置頂" if topic.is_pinned else "取消置頂"
        return jsonify({'message': f'已{status}主題', 'is_pinned': topic.is_pinned})
    except SQLAlchemyError as e:
        db.session.rollback()
        return jsonify({'error': f'操作失敗: {str(e)}'}), 500


@discussion_bp.route('/posts/<int:post_id>/pin', methods=['POST'])
@login_required
def pin_post(post_id):
    """置頂/取消置頂文章 (僅限管理員)"""
    if not session.get('is_admin', False):
        return jsonify({'error': '您沒有管理員權限'}), 403
    
    post = Post.query.get_or_404(post_id)
    
    try:
        post.is_pinned = not post.is_pinned
        db.session.commit()
        
        status = "置頂" if post.is_pinned else "取消置頂"
        return jsonify({'message': f'已{status}文章', 'is_pinned': post.is_pinned})
    except SQLAlchemyError as e:
        db.session.rollback()
        return jsonify({'error': f'操作失敗: {str(e)}'}), 500

File: src/store/modules/discussion.js
import apiService from '@/services/api';

export default {
  namespaced: true,
  
  state: {
    topics: [],
    currentTopic: null,
    posts: [],
    currentPost: null,
    replies: [],
    loading: false,
    error: null,
    pagination: {
      total: 0,
      currentPage: 1,
      totalPages: 1,
    }
  },
  
  mutations: {
    SET_LOADING(state, status) {
      state.loading = status;
    },
    
    SET_ERROR(state, error) {
      state.error = error;
    },
    
    SET_TOPICS(state, { topics, total, pages, currentPage }) {
      state.topics = topics;
      state.pagination.total = total;
      state.pagination.totalPages = pages;
      state.pagination.currentPage = currentPage;
    },
    
    SET_CURRENT_TOPIC(state, topic) {
      state.currentTopic = topic;
    },
    
    SET_POSTS(state, { posts, total, pages, currentPage }) {
      state.posts = posts;
      state.pagination.total = total;
      state.pagination.totalPages = pages;
      state.pagination.currentPage = currentPage;
    },
    
    SET_CURRENT_POST(state, post) {
      state.currentPost = post;
    },
    
    SET_REPLIES(state, { replies, total, pages, currentPage }) {
      state.replies = replies;
      state.pagination.total = total;
      state.pagination.totalPages = pages;
      state.pagination.currentPage = currentPage;
    },
    
    UPDATE_POST_LIKES(state, { postId, likes, likeCount }) {
      const postIndex = state.posts.findIndex(post => post.id === postId);
      if (postIndex !== -1) {
        state.posts[postIndex].likes = likes;
        state.posts[postIndex].like_count = likeCount;
      }
      
      if (state.currentPost && state.currentPost.id === postId) {
        state.currentPost.likes = likes;
        state.currentPost.like_count = likeCount;
      }
    },
    
    UPDATE_REPLY_LIKES(state, { replyId, likes, likeCount }) {
      const replyIndex = state.replies.findIndex(reply => reply.id === replyId);
      if (replyIndex !== -1) {
        state.replies[replyIndex].likes = likes;
        state.replies[replyIndex].like_count = likeCount;
      }
    },
    
    ADD_TOPIC(state, topic) {
      state.topics.unshift(topic);
    },
    
    ADD_POST(state, post) {
      state.posts.push(post);
    },
    
    ADD_REPLY(state, reply) {
      state.replies.push(reply);
    },
    
    UPDATE_TOPIC(state, updatedTopic) {
      const index = state.topics.findIndex(topic => topic.id === updatedTopic.id);
      if (index !== -1) {
        state.topics.splice(index, 1, updatedTopic);
      }
      
      if (state.currentTopic && state.currentTopic.id === updatedTopic.id) {
        state.currentTopic = updatedTopic;
      }
    },
    
    UPDATE_POST(state, updatedPost) {
      const index = state.posts.findIndex(post => post.id === updatedPost.id);
      if (index !== -1) {
        state.posts.splice(index, 1, updatedPost);
      }
      
      if (state.currentPost && state.currentPost.id === updatedPost.id) {
        state.currentPost = updatedPost;
      }
    },
    
    UPDATE_REPLY(state, updatedReply) {
      const index = state.replies.findIndex(reply => reply.id === updatedReply.id);
      if (index !== -1) {
        state.replies.splice(index, 1, updatedReply);
      }
    },
    
    REMOVE_TOPIC(state, topicId) {
      state.topics = state.topics.filter(topic => topic.id !== topicId);
    },
    
    REMOVE_POST(state, postId) {
      state.posts = state.posts.filter(post => post.id !== postId);
    },
    
    REMOVE_REPLY(state, replyId) {
      state.replies = state.replies.filter(reply => reply.id !== replyId);
    }
  },
  
  actions: {
    // 主題相關操作
    async fetchTopics({ commit }, { page = 1, perPage = 10 } = {}) {
      commit('SET_LOADING', true);
      commit('SET_ERROR', null);
      
      try {
        const response = await apiService.discussion.getTopics(page, perPage);
        commit('SET_TOPICS', {
          topics: response.topics,
          total: response.total,
          pages: response.pages,
          currentPage: response.current_page
        });
        return response.topics;
      } catch (error) {
        commit('SET_ERROR', error.message || '獲取主題列表失敗');
        console.error('獲取主題列表失敗', error);
        return [];
      } finally {
        commit('SET_LOADING', false);
      }
    },
    
    async fetchTopic({ commit }, topicId) {
      commit('SET_LOADING', true);
      commit('SET_ERROR', null);
      
      try {
        const response = await apiService.discussion.getTopic(topicId);
        commit('SET_CURRENT_TOPIC', response);
        return response;
      } catch (error) {
        commit('SET_ERROR', error.message || '獲取主題詳情失敗');
        console.error('獲取主題詳情失敗', error);
        return null;
      } finally {
        commit('SET_LOADING', false);
      }
    },
    
    async createTopic({ commit }, data) {
      commit('SET_LOADING', true);
      commit('SET_ERROR', null);
      
      try {
        const response = await apiService.discussion.createTopic(data);
        commit('ADD_TOPIC', response);
        return response;
      } catch (error) {
        commit('SET_ERROR', error.message || '創建主題失敗');
        console.error('創建主題失敗', error);
        throw error;
      } finally {
        commit('SET_LOADING', false);
      }
    },
    
    async updateTopic({ commit }, { topicId, data }) {
      commit('SET_LOADING', true);
      commit('SET_ERROR', null);
      
      try {
        const response = await apiService.discussion.updateTopic(topicId, data);
        commit('UPDATE_TOPIC', response);
        return response;
      } catch (error) {
        commit('SET_ERROR', error.message || '更新主題失敗');
        console.error('更新主題失敗', error);
        throw error;
      } finally {
        commit('SET_LOADING', false);
      }
    },
    
    async deleteTopic({ commit }, topicId) {
      commit('SET_LOADING', true);
      commit('SET_ERROR', null);
      
      try {
        await apiService.discussion.deleteTopic(topicId);
        commit('REMOVE_TOPIC', topicId);
        return true;
      } catch (error) {
        commit('SET_ERROR', error.message || '刪除主題失敗');
        console.error('刪除主題失敗', error);
        throw error;
      } finally {
        commit('SET_LOADING', false);
      }
    },
    
    // 文章相關操作
    async fetchPosts({ commit }, { topicId, page = 1, perPage = 20 }) {
      commit('SET_LOADING', true);
      commit('SET_ERROR', null);
      
      try {
        const response = await apiService.discussion.getPosts(topicId, page, perPage);
        
        if (response.topic) {
          commit('SET_CURRENT_TOPIC', response.topic);
        }
        
        commit('SET_POSTS', {
          posts: response.posts,
          total: response.total,
          pages: response.pages,
          currentPage: response.current_page
        });
        
        return response.posts;
      } catch (error) {
        commit('SET_ERROR', error.message || '獲取文章列表失敗');
        console.error('獲取文章列表失敗', error);
        return [];
      } finally {
        commit('SET_LOADING', false);
      }
    },
    
    async createPost({ commit }, { topicId, data }) {
      commit('SET_LOADING', true);
      commit('SET_ERROR', null);
      
      try {
        const response = await apiService.discussion.createPost(topicId, data);
        commit('ADD_POST', response);
        
        // 如果有當前主題，更新其文章數量
        if (this.state.discussion.currentTopic) {
          const updatedTopic = { 
            ...this.state.discussion.currentTopic,
            post_count: (this.state.discussion.currentTopic.post_count || 0) + 1 
          };
          commit('SET_CURRENT_TOPIC', updatedTopic);
        }
        
        return response;
      } catch (error) {
        commit('SET_ERROR', error.message || '發表文章失敗');
        console.error('發表文章失敗', error);
        throw error;
      } finally {
        commit('SET_LOADING', false);
      }
    },
    
    async updatePost({ commit }, { postId, data }) {
      commit('SET_LOADING', true);
      commit('SET_ERROR', null);
      
      try {
        const response = await apiService.discussion.updatePost(postId, data);
        commit('UPDATE_POST', response);
        return response;
      } catch (error) {
        commit('SET_ERROR', error.message || '更新文章失敗');
        console.error('更新文章失敗', error);
        throw error;
      } finally {
        commit('SET_LOADING', false);
      }
    },
    
    async deletePost({ commit, state }, postId) {
      commit('SET_LOADING', true);
      commit('SET_ERROR', null);
      
      try {
        await apiService.discussion.deletePost(postId);
        commit('REMOVE_POST', postId);
        
        // 如果有當前主題，更新其文章數量
        if (state.currentTopic) {
          const updatedTopic = { 
            ...state.currentTopic,
            post_count: Math.max(0, (state.currentTopic.post_count || 0) - 1)
          };
          commit('SET_CURRENT_TOPIC', updatedTopic);
        }
        
        return true;
      } catch (error) {
        commit('SET_ERROR', error.message || '刪除文章失敗');
        console.error('刪除文章失敗', error);
        throw error;
      } finally {
        commit('SET_LOADING', false);
      }
    },
    
    // 回覆相關操作
    async fetchReplies({ commit }, { postId, page = 1, perPage = 50 }) {
      commit('SET_LOADING', true);
      commit('SET_ERROR', null);
      
      try {
        const response = await apiService.discussion.getReplies(postId, page, perPage);
        
        if (response.post) {
          commit('SET_CURRENT_POST', response.post);
        }
        
        commit('SET_REPLIES', {
          replies: response.replies,
          total: response.total,
          pages: response.pages,
          currentPage: response.current_page
        });
        
        return response.replies;
      } catch (error) {
        commit('SET_ERROR', error.message || '獲取回覆列表失敗');
        console.error('獲取回覆列表失敗', error);
        return [];
      } finally {
        commit('SET_LOADING', false);
      }
    },
    
    async createReply({ commit }, { postId, data }) {
      commit('SET_LOADING', true);
      commit('SET_ERROR', null);
      
      try {
        const response = await apiService.discussion.createReply(postId, data);
        commit('ADD_REPLY', response);
        
        // 如果有當前文章，更新其回覆數量
        if (this.state.discussion.currentPost) {
          const updatedPost = { 
            ...this.state.discussion.currentPost,
            reply_count: (this.state.discussion.currentPost.reply_count || 0) + 1 
          };
          commit('SET_CURRENT_POST', updatedPost);
        }
        
        // 更新文章列表中的回覆數量
        const postsIndex = this.state.discussion.posts.findIndex(post => post.id === postId);
        if (postsIndex !== -1) {
          const updatedPosts = [...this.state.discussion.posts];
          updatedPosts[postsIndex] = {
            ...updatedPosts[postsIndex],
            reply_count: (updatedPosts[postsIndex].reply_count || 0) + 1
          };
          commit('SET_POSTS', {
            posts: updatedPosts,
            total: this.state.discussion.pagination.total,
            pages: this.state.discussion.pagination.totalPages,
            currentPage: this.state.discussion.pagination.currentPage
          });
        }
        
        return response;
      } catch (error) {
        commit('SET_ERROR', error.message || '發表回覆失敗');
        console.error('發表回覆失敗', error);
        throw error;
      } finally {
        commit('SET_LOADING', false);
      }
    },
    
    async updateReply({ commit }, { replyId, data }) {
      commit('SET_LOADING', true);
      commit('SET_ERROR', null);
      
      try {
        const response = await apiService.discussion.updateReply(replyId, data);
        commit('UPDATE_REPLY', response);
        return response;
      } catch (error) {
        commit('SET_ERROR', error.message || '更新回覆失敗');
        console.error('更新回覆失敗', error);
        throw error;
      } finally {
        commit('SET_LOADING', false);
      }
    },
    
    async deleteReply({ commit, state }, replyId) {
      commit('SET_LOADING', true);
      commit('SET_ERROR', null);
      
      try {
        await apiService.discussion.deleteReply(replyId);
        
        // 找到被刪除回覆的文章ID
        const reply = state.replies.find(r => r.id === replyId);
        const postId = reply?.post_id;
        
        commit('REMOVE_REPLY', replyId);
        
        // 如果有當前文章且回覆屬於該文章，更新其回覆數量
        if (state.currentPost && postId === state.currentPost.id) {
          const updatedPost = { 
            ...state.currentPost,
            reply_count: Math.max(0, (state.currentPost.reply_count || 0) - 1)
          };
          commit('SET_CURRENT_POST', updatedPost);
        }
        
        // 更新文章列表中的回覆數量
        if (postId) {
          const postsIndex = state.posts.findIndex(post => post.id === postId);
          if (postsIndex !== -1) {
            const updatedPosts = [...state.posts];
            updatedPosts[postsIndex] = {
              ...updatedPosts[postsIndex],
              reply_count: Math.max(0, (updatedPosts[postsIndex].reply_count || 0) - 1)
            };
            commit('SET_POSTS', {
              posts: updatedPosts,
              total: state.pagination.total,
              pages: state.pagination.totalPages,
              currentPage: state.pagination.currentPage
            });
          }
        }
        
        return true;
      } catch (error) {
        commit('SET_ERROR', error.message || '刪除回覆失敗');
        console.error('刪除回覆失敗', error);
        throw error;
      } finally {
        commit('SET_LOADING', false);
      }
    },
    
    // 讚相關操作
    async likePost({ commit }, postId) {
      try {
        const response = await apiService.discussion.likePost(postId);
        commit('UPDATE_POST_LIKES', {
          postId,
          likes: response.likes,
          likeCount: response.like_count
        });
        return response;
      } catch (error) {
        console.error('操作文章讚失敗', error);
        throw error;
      }
    },
    
    async likeReply({ commit }, replyId) {
      try {
        const response = await apiService.discussion.likeReply(replyId);
        commit('UPDATE_REPLY_LIKES', {
          replyId,
          likes: response.likes,
          likeCount: response.like_count
        });
        return response;
      } catch (error) {
        console.error('操作回覆讚失敗', error);
        throw error;
      }
    },
    
    // 管理員功能
    async pinTopic({ commit }, topicId) {
      try {
        const response = await apiService.discussion.pinTopic(topicId);
        
        // 更新主題列表中的置頂狀態
        const topics = [...this.state.discussion.topics];
        const index = topics.findIndex(t => t.id === topicId);
        
        if (index !== -1) {
          topics[index] = { ...topics[index], is_pinned: response.is_pinned };
          
          // 如果狀態改變，可能需要重新排序
          if (response.is_pinned) {
            // 移到最前面
            const pinnedTopic = topics.splice(index, 1)[0];
            topics.unshift(pinnedTopic);
          }
          
          commit('SET_TOPICS', {
            topics,
            total: this.state.discussion.pagination.total,
            pages: this.state.discussion.pagination.totalPages,
            currentPage: this.state.discussion.pagination.currentPage
          });
        }
        
        // 如果是當前主題，也要更新
        if (this.state.discussion.currentTopic && this.state.discussion.currentTopic.id === topicId) {
          commit('SET_CURRENT_TOPIC', {
            ...this.state.discussion.currentTopic,
            is_pinned: response.is_pinned
          });
        }
        
        return response;
      } catch (error) {
        console.error('置頂/取消置頂主題失敗', error);
        throw error;
      }
    },
    
    async pinPost({ commit }, postId) {
      try {
        const response = await apiService.discussion.pinPost(postId);
        
        // 更新文章列表中的置頂狀態
        const posts = [...this.state.discussion.posts];
        const index = posts.findIndex(p => p.id === postId);
        
        if (index !== -1) {
          posts[index] = { ...posts[index], is_pinned: response.is_pinned };
          
          // 如果狀態改變，可能需要重新排序
          if (response.is_pinned) {
            // 移到最前面
            const pinnedPost = posts.splice(index, 1)[0];
            posts.unshift(pinnedPost);
          }
          
          commit('SET_POSTS', {
            posts,
            total: this.state.discussion.pagination.total,
            pages: this.state.discussion.pagination.totalPages,
            currentPage: this.state.discussion.pagination.currentPage
          });
        }
        
        // 如果是當前文章，也要更新
        if (this.state.discussion.currentPost && this.state.discussion.currentPost.id === postId) {
          commit('SET_CURRENT_POST', {
            ...this.state.discussion.currentPost,
            is_pinned: response.is_pinned
          });
        }
        
        return response;
      } catch (error) {
        console.error('置頂/取消置頂文章失敗', error);
        throw error;
      }
    }
  },
  
  getters: {
    isLoading: state => state.loading,
    error: state => state.error,
    allTopics: state => state.topics,
    getTopic: state => topicId => state.topics.find(t => t.id === topicId),
    currentTopic: state => state.currentTopic,
    allPosts: state => state.posts,
    getPost: state => postId => state.posts.find(p => p.id === postId),
    currentPost: state => state.currentPost,
    allReplies: state => state.replies,
    getReply: state => replyId => state.replies.find(r => r.id === replyId),
    pinnedTopics: state => state.topics.filter(t => t.is_pinned),
    regularTopics: state => state.topics.filter(t => !t.is_pinned),
    pinnedPosts: state => state.posts.filter(p => p.is_pinned),
    regularPosts: state => state.posts.filter(p => !p.is_pinned),
    pagination: state => state.pagination
  }
};